src/main/java/at/ac/tuwien/kr/alpha/grounder/Substitution.java

```diff
@@ -68,29 +75,98 @@ public static Substitution unify(Literal literal, Instance instance, Substitutio
 	}
 
 	/**
-	 * Computes the unifier of the atom and the instance and stores it in the variable substitution.
-	 * 
-	 * @param atom         the body atom to unify
-	 * @param instance     the ground instance
-	 * @param substitution if the atom does not unify, this is left unchanged.
-	 * @return true if the atom and the instance unify. False otherwise
+	 * Helper class to lazily clone the input substitution of Substitution.unify only when needed.
 	 */
-	public static Substitution unify(Atom atom, Instance instance, Substitution substitution) {
-		for (int i = 0; i < instance.terms.size(); i++) {
-			if (instance.terms.get(i) == atom.getTerms().get(i) || substitution.unifyTerms(atom.getTerms().get(i), instance.terms.get(i))) {
-				continue;
+	private static class UnificationHelper {
+		Substitution returnSubstitution;
+
+		Substitution unify(List<Term> termList, Instance instance, Substitution partialSubstitution) {
```

Comment:
Note that this is a matter of personal taste, but I'm not sure I agree with the return types here..
I think it might make sense if `unifyTerms` returned something other than boolean, thereby getting rid of the side-effects of potentially modifying the given partialSubstitution. Instead, `unifyTerms` could return a `Substitution` that may be either a new (cloned) one, or the one passed in in case nothing changed. Alternatively, we could also use an `Optional<Substitution>` here, where the presence of a value would indicate that a new variable got bound. The version with `Optional` could also be used for `Substitution::unify` as an alternative to returning `null`.

Reply:
Letting `unifyTerms` return a `Substitution` is a good idea. I would not want to return an `Optional` however, since substitutions are used at the core of grounding and `Optional` is likely slower and creates more objects.

Reply:
Actually, I decided to remove `unifyTerms` as it was only ever used in unit tests. These unit tests have been adapted to now test the new/renamed `specializeSubstitution`.